### PR TITLE
alloc: Add unstable Cfg feature `no_global_oom_handling

### PR DESCRIPTION
**Pull request by Ericson2314**
For certain sorts of systems, programming, it's deemed essential that
all allocation failures be explicitly handled where they occur. For
example, see Linus Torvald's opinion in [1]. Merely not calling global
panic handlers, or always `try_reserving` first (for vectors), is not
deemed good enough, because the mere presence of the global OOM handlers
is burdens static analysis.

One option for these projects to use rust would just be to skip `alloc`,
rolling their own allocation abstractions.  But this would, in my
opinion be a real shame. `alloc` has a few `try_*` methods already, and
we could easily have more. Features like custom allocator support also
demonstrate and existing to support diverse use-cases with the same
abstractions.

A natural way to add such a feature flag would a Cargo feature, but
there are currently uncertainties around how std library crate's Cargo
features may or not be stable, so to avoid any risk of stabilizing by
mistake we are going with a more low-level "raw cfg" token, which
cannot be interacted with via Cargo alone.

Note also that since there is no notion of "default cfg tokens" outside
of Cargo features, we have to invert the condition from
`global_oom_handling` to to `not(no_global_oom_handling)`. This breaks
the monotonicity that would be important for a Cargo feature (i.e.
turning on more features should never break compatibility), but it
doesn't matter for raw cfg tokens which are not intended to be
"constraint solved" by Cargo or anything else.

To support this use-case we create a new feature, "global-oom-handling",
on by default, and put the global OOM handler infra and everything else
it that depends on it behind it. By default, nothing is changed, but
users concerned about global handling can make sure it is disabled, and
be confident that all OOM handling is local and explicit.

For this first iteration, non-flat collections are outright disabled.
`Vec` and `String` don't yet have `try_*` allocation methods, but are
kept anyways since they can be oom-safely created "from parts", and we
hope to add those `try_` methods in the future.

[1]: https://lore.kernel.org/lkml/CAHk-=wh_sNLoz84AUUzuqXEsYH35u=8HV3vK-jbRbJ_B-JjGrg@mail.gmail.com/

### CODE REVIEW

**Comment by Ericson2314** [comment]
I added the WIP because something in `alloc` is requiring `exchange_malloc`, yet I don't see any `box` expressions except for in tests, and when I dumped MIR for the module I didn't see it used.

**Comment by Ericson2314** [comment]
Also not sure why that failure happened, because I don't see any `default-features = false` or similar for alloc.

**Comment by bjorn3** [comment]
> I added the WIP because something in `alloc` is requiring `exchange_malloc`, yet I don't see any `box` expressions except for in tests, and when I dumped MIR for the module I didn't see it used.

`Box::new` needs it.

**Comment by the8472** [comment]
This is a lot of `cfg` annotations. Would it make sense to group functions those into internal modules instead? E.g. `vec::fallible` and `vec::infallible` both containing `impl` blocks for the same types and then putting the `cfg` on the whole module.

Of course that'll require a lot of code motion now, but but it provides clearer separation and it'll be more obvious in the future when people add things in the wrong place.

**Comment by phil-opp** [comment]
> Also not sure why that failure happened, because I don't see any `default-features = false` or similar for alloc.

The test seems to [use `rustc` directly](https://github.com/rust-lang/rust/blob/080d30235fff4373fc5ce9f84875e3be91d1a09e/src/test/run-make-fulldeps/alloc-extern-crates/Makefile). I think that (default) features are normally handled by cargo and [only passed to rustc as `--cfg` argument](https://github.com/rust-lang/cargo/blob/163097cdd9da796b2e50f983343ea451aa6a3e91/src/cargo/core/compiler/mod.rs#L880-L882), so we probably need to adjust the test to enable the `global_oom_handling` feature explicitly.

**Comment by Aaron1011** [comment]
What happens if someone forgets to put `#[cfg(feature = "global-oom-handling")]` on an API that depends on global OOM handling? Will compilation fail with the feature disabled, or will the user get a weird runtime problem?

**Comment by bjorn3** [comment]
It will fail with a compilation error as `handle_alloc_error` is cfg'ed away.

**Comment by Ericson2314** [comment]
Indeed that is what's happening with right now.

I should have those 2 things fixed in a bit. Thanks for the tips, everyone.

**Comment by Ericson2314** [comment]
@the8472 It is a lot of `cfg`, but I think the "boundaries" between e.g. `vec::fallible` and `vec::infallible` would change a lot as more `try_*` methods are created. That sort of churn will cause massive headaches with PR conflicts, so I think gratuitous `cfg` is in fact the far lesser evil, at least until we decide how many `try_*` methods we want and stabilize around that.

**Comment by Ericson2314** [comment]
The PR is passing, but it would be nice to have a test that the build with `global-oom-handling` *disabled* doesn't break. How ought I to do that?

File: src/test/run-make-fulldeps/alloc-extern-crates/Makefile [hunk 1]
```diff
@@ -2,4 +2,4 @@
 
 all:
 	$(RUSTC) fakealloc.rs
-	$(RUSTC) --edition=2018 --crate-type=rlib ../../../../library/alloc/src/lib.rs --cfg feature=\"external_crate\" --extern external=$(TMPDIR)/$(shell $(RUSTC) --print file-names fakealloc.rs)
+	$(RUSTC) --edition=2018 --crate-type=rlib ../../../../library/alloc/src/lib.rs --cfg feature=\"external_crate\" --cfg=feature=\"global-oom-handling\" --extern external=$(TMPDIR)/$(shell $(RUSTC) --print file-names fakealloc.rs)
```
Comment (bjorn3) [inline src/test/run-make-fulldeps/alloc-extern-crates/Makefile:5]: This test doesn't seem to test anything anymore: https://github.com/rust-lang/rust/search?q=external_crate

Comment (bjorn3) [reply src/test/run-make-fulldeps/alloc-extern-crates/Makefile:5]: By the way I think you can use the same pattern to test liballoc without the `global-oom-handling` feature.

Comment (Ericson2314) [reply src/test/run-make-fulldeps/alloc-extern-crates/Makefile:5]: Agreed it does seem that way.

**Comment by the8472** [comment]
> @the8472 It is a lot of `cfg`, but I think the "boundaries" between e.g. `vec::fallible` and `vec::infallible` would change a lot as more `try_*` methods are created. That sort of churn will cause massive headaches with PR conflicts, so I think gratuitous `cfg` is in fact the far lesser evil, at least until we decide how many `try_*` methods we want and stabilize around that.

I'm not sure what you're concerned about. Potential churn in the long-term once those modules are established or the way to creating those modules?

If it's about the latter then we can probably minimize the conflict potential by only creating a `fallible` module and moving the `try_` code and its fallible callees there. That should be a fairly small API surface right now and thus not need too much code movement. Once that is done we only need to slap a `cfg` on top of the `impl` blocks which contain infallible APIs to disable them.

**Comment by Ericson2314** [comment]
> I'm not sure what you're concerned about. Potential churn in the long-term once those modules are established or the way to creating those modules?

Churn in the short term why `try_` methods are being created.

> If it's about the latter then we can probably minimize the conflict potential by only creating a `fallible` module and moving the `try_` code and its fallible callees there. That should be a fairly small API surface right now and thus not need too much code movement. Once that is done we only need to slap a `cfg` on top of the `impl` blocks which contain infallible APIs to disable them.

When a `try_` method is made, usually what happens is the existing implementation is refactored into the `try_` version, and then the old one becomes a small wrapper around that. So with even just `fallible` module, we have a bunch of churn when the implementation gets refactored *and moved* from one module too there. And git handles refactor + moves extremely poorly.

**Comment by the8472** [comment]
> Churn in the short term why `try_` methods are being created.

I don't understand what you mean by that. The try methods have to be created either way. They're not just created for a short term reason and then removed later. So "churn in the short term" cannot be their raison d'être.

> When a `try_` method is made, usually what happens is the existing implementation is refactored into the `try_` version, and then the old one becomes a small wrapper around that. So with even just `fallible` module, we have a bunch of churn when the implementation gets refactored _and moved_ from the original module to `fallible` one. And git handles refactor + moves extremely poorly.

I don't see the difference here. You need to create a new method which is sortof but not quite copy-pasted from the old one. And then redirect the old one to the new one. This has to happen either way whether they live in different files or not and whether you annotate on the function level or not. And if necessary can be split into multiple commits to make it easier for git to follow the movement.

E.g. you could implement the try method first in the base module and then once it is done use a separate commit to move it into the infallible module.

Yes, in the short-term this means a little more work, but in the long-term it makes it clearer which part belongs to the restricted fallible API set and which to the more lax infallible APIs.

**Comment by Ericson2314** [comment]
Thanks @bjorn3 for laying this out much more clearly than I would have.

**Comment by the8472** [comment]
> Git blame doesn't track moves between files at all unless you pass it `-C`. The github ui doesn't have any option to enable this and most people likely don't know about it at all.

I assume you're using that already, otherwise the occasional refactor that splits large modules into smaller ones (as happend not too long ago with `Vec`) would cause the very same issues.
And since there's no ban on such split operations (in fact they're encouraged by tidy's file length limit) why not do another one to isolate the fallible methods?

> When doing it within the same file, git will consider most lines of non-try function in the previous version to be part of the try function in the current version. 

Yes, hence do that in one commit, then do the motion later so that git can follow it.

**Comment by Ericson2314** [comment]
@the8472 I'm not against ever rearranging code --- quite in support in fact. But I imagine the `try_` methods will be added incrementally over a long period of time. (We could do them all at once, but I'm afraid it's rather controversial.)

Assuming I'm correct in saying there will be many commits adding `try_` methods, I rather have one big reorg at the end, then constant reorging, because it's much less a burden on others.

**Comment by the8472** [comment]
Ok, that might work. But on the other hand the current approach also means people have to reason about fallibility on a per-method level. And if methods are added incrementally that would also keep church limited in impact, so moving wouldn't be that bad either?

**Comment by Ericson2314** [comment]
> But on the other hand the current approach also means people have to reason about fallibility on a per-method level.

Well, I hope the `try_` prefix helps with that.

> And if methods are added incrementally that would also keep church limited in impact, so moving wouldn't be that bad either?

You mean entirely new methods that don't have a `try_` or non-`try_` version today?

**Comment by the8472** [comment]
> > But on the other hand the current approach also means people have to reason about fallibility on a per-method level.
> 
> Well, I hope the `try_` prefix helps with that.

Except that the cfg has to be added to anything that is infallible. So when you're not working on fallible code you still have to keep the cfg that was introduced for fallible APIs in mind.

It even affects internal helper methods because those would be dead code or fail to build if the public API were cfg'd off. So this is quite infectious.

If it could be put on top of impl blocks and modules instead of individual functions and the fallible APIs moved out of those blocks/modules then this would cause fewer build failures down the road for those working on infallible functions.

> > And if methods are added incrementally that would also keep church limited in impact, so moving wouldn't be that bad either?
> 
> You mean entirely new methods that don't have a `try_` or non-`try_` version today?

No, I mean if the `try_` methods are added incrementally as you said in your previous post then adding *and* moving them shouldn't be all that disruptive since it's only done in small increments so there won't be as many PRs in flight that might conflict with them.
Other than the initial move of those that already are fallible, but there aren't that many yet.

**Comment by nbdd0121** [comment]
I agree with @the8472 that we should separate `fallible` and `infallible`. `cfg` each method is quite fragile to breakage: people can accidentally forget to `cfg` or call a infallible method from a supposedly fallible method. Having them in separate places would be helpful so that adding methods would require you to be aware of whether the method is fallible or not first, and we wouldn't have to worry about forgetting to `cfg` because all infallible methods could be turned off together.

**Comment by Ericson2314** [comment]
I'm not worried about people forgetting because we'll have CI. Ultimately the choice is not mine; I've made my arguments and will go with whatever the reviewers want.

**Comment by tschuett** [comment]
Do you want to write an RFC documenting the future direction for local allocation error handling?

**Comment by Ericson2314** [comment]
I could I suuppose, but to be honest I was hoping this could fall under the preview of the allocators WG (just like the exciting `try_` methods) and not require jumping through so many hoops. As long as `-Z build-std` is unstable, I don't think there are any stability guarantees around Cargo features in stdlib crates.

**Comment by tschuett** [comment]
No worries. I was hoping for more attention for your challenges.

File: library/alloc/src/rc.rs [hunk 2]
```diff
@@ -255,19 +255,27 @@ use core::convert::{From, TryFrom};
 use core::fmt;
 use core::hash::{Hash, Hasher};
 use core::intrinsics::abort;
+#[cfg(feature = "global-oom-handling")]
```
Comment (Mark-Simulacrum) [inline library/alloc/src/rc.rs:258]: I assume this is being added (as well as some other import cfgs) due to being unused after cfgs later in this file? Could we instead add allow(unused) or do, gated at crate root, enabled when the feature is off?

**Comment by Mark-Simulacrum** [comment]
I think this may not be the strategy we end up with, but it doesn't seem entirely unreasonable to start out with. I left a comment which, if applied, would likely help cut down on the number of changes necessary (I hope) to make this happen, particularly around imports.

Cc @rust-lang/libs 

I do think this is likely to be a pretty painful continuous burden as new changes are added; it may be a good idea to have some discussion over the preferred strategy here. If it's try_ methods, this might be a good option; otherwise, it seems plausible that some other point in the design space may be better.

**Comment by Ericson2314** [comment]
@Mark-Simulacrum Thanks for being willing to give this a shot. People have long been uncomfortable with `try_` methods, but I haven't seen counterproposals that aren't much far involved (`result<T, !> -> T` coercions, for example). I think there is a fairly imminent need for something like this in the Linux work, and if we eventually find another approach refactoring this to that shouldn't be too hard.

**Comment by SimonSapin** [comment]
> I think this may not be the strategy we end up with, but it doesn't seem entirely unreasonable to start out with.

This reasoning applies to merging an unstable functionality. However we don’t have a mechanism to make a Cargo `[feature]` only available on Nightly, nor making *disabling* a default feature only possible on Nightly. (How would that even work?).

Cargo `[features]` of standard library crates are only relevant when (re)compiling these crates, which currently is only possible on Nightly because Cargo’s `-Z build-std` is still unstable. But at some point `-Z build-std` will (hopefully) be stabilized. At that point are we automatically committing at the same time to forever keep the ability to build an `alloc` crate without `handle_alloc_error`?

**Comment by Ericson2314** [comment]
@SimonSapin It's a good point. I think we simply should make a note to revisit this when it comes time to stabilize `-Z build-std` so we don't accidentally stabilize it, and deal with it between now and then. Anything else is going to hold this up indefinitely and jeopardize Rust in Linux.

I personally think `-Z build-std` is a great idea implemented in a really wrong way, so I am not too concerned with it being stabilized any time soon.

**Comment by Mark-Simulacrum** [comment]
I think any stabilization would need to tackle that question (either for build-std or some other mechanism, such as shipping alloc with this feature disabled for some target), but I don't think we need to explicitly say no for now just because a mechanism is not readily available.

I think, for example, we could hide alloc behind a blanket re-export shim which doesn't expose this feature.

**Comment by SimonSapin** [comment]
> I think we simply should make a note to revisit this when it comes time to stabilize -Z build-std so we don't accidentally stabilize it, and deal with it between now and then. Anything else is going to hold this up indefinitely and jeopardize Rust in Linux.

The same argument can work the other way around: the timing might end up being that the "just make a note" strategy holds up `-Z build-std` indefinitely and jeopardizes it.

**Comment by tschuett** [comment]
As an outsider, I have the feeling that this not the right forum to discuss the mechanisms for Rust for Linux.

**Comment by the8472** [comment]
Could some `cfg` trickery be used to make the build fail if std is built without this feature on a non-nightly channel?

**Comment by Ericson2314** [comment]
> > I think we simply should make a note to revisit this when it comes time to stabilize -Z build-std so we don't accidentally stabilize it, and deal with it between now and then. Anything else is going to hold this up indefinitely and jeopardize Rust in Linux.
> 
> The same argument can work the other way around: the timing might end up being that the "just make a note" strategy holds up `-Z build-std` indefinitely and jeopardizes it.

Sure, in theory, but `-Z build-std` is massively flawed so if we're going to fix that it will be far more work than teaching cargo some notion of unstable features that only the stdlib is allowed to used. So I don't think it will come up with practice

Also, the current features are quite ad-hoc (hacks for compiler-builtins, etc.) which makes me think at no point were the stdldib features being groomed for posterity. I don't think we need hold this to a higher standard than those if they all will alike block `-Z build-std`.

**Comment by SimonSapin** [comment]
Separately from the API stability concern above, I’d like @rust-lang/libs to discuss whether to pursue this at all. Should the standard library’s `alloc` crate support this functionality in the first place?

This PR’s diff and discussion thread show that it introduces non-trivial maintenance burden. `#[cfg]`s are added everywhere, new CI is needed make sure this does not regress, and all most future changes to `alloc` will need to account to for this other configuration in some way to avoid breaking it. Likely by adding more `#[cfg]`, possibly with contributors only realizing they need to after a first round of CI fails.

The use case is valid, I’m not disputing that, but it is also extremely niche.

Correctly recovering from allocation failure in all cases is [reportedly](https://twitter.com/fitzgen/status/1383455503938772998) extremely difficult. Having containers with (only) fallible `try_*` APIs is only a start: what should the project’s own code do in all of these error code paths? (This contributes to most projects not trying, making the use case more niche.) On userland Linux, over-commit mean that allocation almost never fails in practice. Especially on 64-bit where you pretty much have to actively try before the address space can be exhausted.

Just because a use case exists doesn’t mean that the standard library *must* support it. https://github.com/Rust-for-Linux/linux/issues/2 even says:

> It is likely we will end up designing our own `alloc` crate

… which would make this effort irrelevant to the Linux kernel. This PR’s description says about that outcome:

> But this would, in my opinion be a real shame.

I understand that sentiment, but avoiding that "real shame" is a weak argument in my opinion to justify this maintenance burden.

**Comment by SimonSapin** [comment]
I’d really like to see Rust-for-Linux succeed. For that I think it is important to minimize on the critical path the number of changes that need to happen in upstream Rust, especially ones that are disruptive or involve design that will take a long time to stabilize.

I tried and failed to stabilize https://github.com/rust-lang/rust/issues/48043 *two years ago*. And that’s only `try_reserve`! This PR is based on the premise of designing, adding, and stabilizing fallible `try_*` APIs for *every* functionality in `alloc` that relevant projects may want to use.

Having Rust-for-Linux fork the `alloc` crate or write their own containers is comparatively much more realistic in my opinion, if hoping to ship months from now rather than years from now.

**Comment by yaahc** [comment]
Regarding @Mark-Simulacrum and @SimonSapin's suggestions that the libs team discuss this I've gone ahead and nominated the issue for discussion in our next meeting ([calendar if you need it](https://github.com/rust-lang/libs-team#quick-facts)). Also, Mark has gone ahead and created zulip topic for discussion in the mean time https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/alloc.20non-global-oom.20support.

**Comment by joshtriplett** [comment]
I think fallible allocation methods are useful for more than just the Linux kernel. Firmware may want that as well, as might low-level userspace software (e.g. init implementations) or some libraries. I'd love to have these methods available in general, and have a way to enforce that the panic-on-oom versions don't exist and can't be called.

I think this approach makes some sense. I also think, before build-std is stabilized, we're going to want a version of this that works for code that doesn't build its own copy of std/alloc, and just prevents calling the methods (and thus lets them be omitted from the final binary).

**Comment by ojeda** [comment]
At the moment, I am favoring a custom `alloc` crate mainly due of the time frame issue. At Rust for Linux we do not have the same restrictions as Rust: we can tweak any and all internal APIs as we please, without backwards compatibility concerns. By contrast, Rust needs to get `alloc` right (module editions, deprecation notices, etc.); and doing that to the point of stabilization seems like it will take a long time.

On way or the other, we hope that whatever we end up doing in Rust for Linux informs the design of `alloc`. In fact, even if we go for a custom crate, we could go back later to reuse `alloc` when the needed parts are stabilized (assuming they fit our needs etc.).

**Comment by c410-f3r** [comment]
cc https://github.com/c410-f3r/rfcs/blob/mem_structs/text/2978-stack_based_vec.md#verifiable-methods

**Comment by Ericson2314** [comment]
Wow, a bunch of comments as I was responding, partially mimicking what I said, but I'll post this anyways.

> The use case is valid, I’m not disputing that, but it is also extremely niche.

So while Linus is probably the most high-profile person complaining about this, it's not just Rust for Linux: People have griped about the rust stdlib's OOM policy as unbecoming for a systems language since Rust was created. I'd love to finally refute that line of criticism once and for all.

> Correctly recovering from allocation failure in all cases is reportedly extremely difficult.

This is really important: `try_` is *not* just for people that want to make all allocation failures recoverable. It's also for those that want to be able to audit the allocations failures which aren't recovered from. Right now, if one wants to have such a whitelist --- too bad. There's no good way to do the analysis.

> I understand that sentiment, but avoiding that "real shame" is a weak argument in my opinion to justify this maintenance burden.

I don't think this is an apples to oranges "niceties" comparison. The cost of having a gazillion project specific allocation libraries is far less code reuse and library interopt. Now the stdlib developers don't pay that cost, but the Rust community does, and it's far more time/efficiency loss than dealing with these CFGs.

**Comment by kornelski** [comment]
I am concerned that a global absolute switch will be impractically hard to enable. If it's all-or-nothing, it means *all* crates in the entire dependency tree need to cooperate, and *every* function they have must be compatible. I'm afraid that it would be a struggle to convince everyone to support fallible alloc everywhere, even in places where it really doesn't matter at all.

For example, I'm running a complex server in a memory-constrained environment, which seriously needs to handle OOM in its request handlers. But the server also has a bunch of startup code executed only once, and OOM is absolutely not a concern there (i.e. if it ever OOMed at such an early stage, it must have been due to hopelessly broken situation, and abort would be the only option anyway). 

And my startup code uses an argument parsing library. I wouldn't want to tell authors of that argument parsing library that they should rewrite their library to fully support fallible OOM handling everywhere, even though realistically it will never ever happen in their code and nobody will ever care about OOM in an arg parsing library. 

Enforcement of fallible allocations is not an application-global problem, even in applications that do care about OOM.

**Comment by alex** [comment]
While I think it's true that not every use case that cares about fallible allocations wants "absolutely no global oom under any circumstances" (Firefox is in this category), I think there's also cases that do want this.

Which to me means that while option can't be the _only_ part of the fallible allocation story, I don't think the existence of other use cases means we _shouldn't_ do this.

**Comment by Ericson2314** [comment]
@alex is exactly right: they both uses are legit, and we can have solutions for both.

@kornelskI I've earlier proposed some more elaborate designs of leveraging an associated error type on allocators to gate whether the non-`try_` methods could be called. But this has been a big thing to coordinate, at least early in the design processes of the allocators API, being a more complicated design motivated by a more nuanced problem.

I think both solutions are complimentary, and if we start with this for the "linuxes", then stuff for the "firefoxes" will be easier to discuss later.

**Comment by the8472** [comment]
For the linux use-case would it be possible to leave the offending functions in std but check that they don't end up in the final binary via CI, e.g. by using  https://github.com/philipc/findpanics (not sure if that would work in kernel modules).

Then we could focus on adding the necessary fallible methods and making it more ergonomic to avoid the infallible would be a separate task.

**Comment by Ericson2314** [comment]
While I was dismissive of splittings modules above, spitting *crates*, like having  `alloc` depend on `alloc-no-global-ooom-handler` or whatever, however does have a good benefit in that we don't need any Cargo features or extra CI, since crate boundaries are already enforced.

Now the churn argument might still indicate we should start with features and cfgs, but if we did reach a steady point where we had finished writing `try_` methods but Cargo features were still annoying to test combinatorially, we might go then go straight to the above.

**Comment by the8472** [comment]
I'm not sure that would be so easy. Splitting crates would either require some compiler magic to allow inherent implementations for those types to exist in two crates or at least some `#[path]` hacks to "inline" the crate rather than importing it (as was done for `std_detect` until recently and made some `use` declarations inside std more unwieldy).

**Comment by SimonSapin** [comment]
A separate crate can’t add inherent methods to types, though. It can create a new type that forwards the upstream methods (which allows adding more methods locally), but then you get two types that can’t be used interchangeably (and it’s a lot of duplication of signatures and doc-comments).

I don’t know if turning inherent methods to methods on extension traits in the prelude could cause breakage. But either way, the `alloc` crate doesn’t have a prelude that gets implicitly imported by `no_std` crates,.

**Comment by Ericson2314** [comment]
> I don’t know if turning inherent methods to methods on extension traits in the prelude could cause breakage.

Yes only that way makes sense. I don't know whether it would either.

**Comment by yaahc** [comment]
> I don’t know if turning inherent methods to methods on extension traits in the prelude could cause breakage. 

I would expect this to cause ambiguous method resolution errors if traits with the same method name already exist, though this seems like the type of breakage we'd allow if it goes through crater cleanly.

**Comment by Ericson2314** [comment]
@yaahc Because every new extension trait method was formerly an inherent method, I think there will be no new conflicts? Someone could import the `alloc-no-global-ooom-handler` crate alone, make their own non-`try_` methods and then get ambiguity errors if they then use `alloc`, but that's on them as far as I'm concerned --- not fundamentally different from making your own `Vec` today and also getting ambiguity errors.

**Comment by Ericson2314** [comment]
Based on discussion with the last 2 lib team meetings, I've changed this to use "raw" CFG instead of Cargo features. I was going to write more details, but i don't want to give any false impression that I, as a mere guest to the meetings, was giving an authoritative summary. But I can say this is an outcome that I am happy with: it unblocks things now, and keeps the door plenty open to alternative designs later.

**Comment by Ericson2314** [comment]
(Rebased in hopes that there isn't another spurious network error in the MinGW build.)

**Comment by m-ou-se** [comment]
@Ericson2314 Thanks again for taking the time to join our meetings, and your patience with our long discussions! :)

For everyone who was not in the meeting, these were our conclusions:

1. We are *not* comfortable with using `cfg(feature = ..)` for this, since we don't know yet what enabled-by-default features will mean for stability. Right now everything involving features is unstable, but once we get build-std closer to stabilization, it's might be tricky to deal with existing enabled-by-default features.
2. We *do* feel comfortable with `cfg(something)`, since that's more clearly unstable and does not carry any expectations that features have or might have in the future.
3. We do want the CI to at least run the equivalent of 'cargo check' with this cfg, to make sure it stays in a buildable state.
4. These `cfg` attributes can be a placeholder for a better feature that might come along in the future. For example, a 'warning, this might panic due to no fault of the caller'-attribute, or 'might-oom' attribute, or a `cfg(feature = ..)` attribute, or somethinge else. They are trivial to grep for and remove or replace.
5. We don't know yet what new OOM-safe APIs for the std collections should look like. Maybe separate types (Vec and oomsafe::Vec?) or just new methods on the existing types (Vec::try_reserve, etc.). Maybe even some very minimal interface that a community crate can provide an interface upon. This will require a concrete proposal in an RFC that explains the benefits and downsides, and can be discussed then.

So, as @Mark-Simulacrum already said:

> I think this may not be the strategy we end up with, but it doesn't seem entirely unreasonable to start out with.

**Comment by Mark-Simulacrum** [comment]
This looks pretty good to me. I haven't done a thorough review of the changed APIs, but presuming that CI passes and there's no complaints on nightly or so, I don't expect we need to be too careful: this can at worst hide something that shouldn't be, and we'll likely notice pretty quickly.

I did have a passing thought that maybe the cfg's should be all a cfg(accessible) of some function (e.g., handle_alloc_error), and then we would be able to apply more fancy logic to that functions existence. But this need not (and should not) block this PR.

So, in that light, @bors r+ rollup=never

**Comment by ojeda** [comment]
> We don't know yet what new OOM-safe APIs for the std collections should look like. Maybe separate types (Vec and oomsafe::Vec?) or just new methods on the existing types (Vec::try_reserve, etc.). Maybe even some very minimal interface that a community crate can provide an interface upon. This will require a concrete proposal in an RFC that explains the benefits and downsides, and can be discussed then.

I am very happy to hear this! As I mentioned above, taking enough time to get standard `alloc` right it seems like the best approach.

And having a `cfg` flag seems like a good first step that can reasonably be taken right away and is likely to be quite useful to projects out there.

Thanks for this!

**Comment by TimDiekmann** [comment]
Is it intended, that `Arc::new` and `Rc::new` were not gated? They use the `box` syntax so the test does not cover it.

**Comment by SimonSapin** [comment]
What happens today when using `box` syntax in a `no_std` program without the `alloc` crate? Is it different in a `no_std` program with `alloc` but without global `handle_alloc_error`?

**Comment by bjorn3** [comment]
```
$ echo '#![feature(box_syntax)] #![no_std] fn foo() { let _ = box 42; }' | rustc +nightly - --crate-type lib
error: requires `owned_box` lang_item

error: aborting due to previous error
```

**Comment by bjorn3** [comment]
To simulate liballoc without global `handler_alloc_error`:
```
$ echo '#![feature(box_syntax, lang_items)] #![no_std] #[lang = "owned_box"] struct Box<T: ?Sized>(*mut T); pub fn foo() { let _ = box 42; }' | rustc +nightly - --crate-type lib
error: requires `exchange_malloc` lang_item

error: aborting due to previous error
```
(requires `foo` to be codegened for an error)

**Comment by Ericson2314** [comment]
@TimDiekmann I did not include them in error. I just added cfg until it worked; pity `box` leads only to an error post-monomorphization.

**Comment by safinaskar** [comment]
Important news! `HashMap` got method `try_insert` ( https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.try_insert ), where `try_` means something completely different to `Box::try_new`. Please, prevent somehow this `try_insert` from becoming stable

**Pull request merged**